verified_boot_app.md

```diff
@@ -57,12 +173,112 @@ Src: https://github.com/Kicksecure/kicksecure-wiki-backup/blob/6900fdfbdea8693fd
 ** Power on -> firmware detects an "augmented" Linux distribution such as Kicksecure with a key database in the well-known location -> show key from key database to the user -> user confirms or denies the key -> continue boot (if confirmed, otherwise go back to firmware menu or something)
 ** User disables Secure Boot -> as usual
-* Additional notes:
-** To handle the situation where multiple shims are installed, the firmware may need the ability to find multiple candidate bootloaders in the removable media path (or any bootloader path)
+-->
 
 -->
+
+### Initial configuration
+
+<!-- TODO: diagrams? -->
+
+1. User powers on the platform.
+2. Firmware performs boot media enumeration.
+3. Firmware detects initial boot or boot with default settings and launches an
+   application.
+4. The application visits all handles with `gEfiSimpleFileSystemProtocolGuid`
+   and `gEfiPartTypeSystemPartGuid`.
+5. For each handle, find operating system bootloader. E.g. GRUB, SHIM, Windows
+   Boot Manager.
+6. For each handle, search for files containing keys and certificates. E.g.
+   files with `.der`, `.cer`, `.key` extensions.
+7. For each bootloader found in step 5, match a key found in step 6 or
+   `PKDefault`, `KEKDefault`, `dbDefault`, `dbxDefault`.
+8. If key not in `dbxDefault`, attempt image verification and report the
+   result to the user.
+9. If verification passed, ask the user to trust the key and optionally offer
+   to add a boot option.
+11. If the user decides to trust the key, the key is added to `DB`.
+12. Ask the user to continue boot or proceed with next bootloader verification
+    (go to step 8).
+13. If the user decided to continue boot, ask the user to enroll a `PK` from a
+    pool of detected keys. Then enable Secure Boot and proceed with booting.
+14. If no more bootloaders present and user did not trust any key, go to the
+    firmware setup menu.
+
+### Secure Boot fails to verify the image
+
+1. User powers on the platform.
+2. Firmware performs boot media enumeration.
+3. Firmware attempts to boot a boot option according to boot order.
+4. Image pointed by the boot option fails the verification.
+5. The firmware launches the application to handle possibly new operating
+   system or a change in the operating system keys/components.
+6. The application scans the partitions for new and untrusted keys.
+7. The application verifies if the boot options in the boot order have a
+   matching key and pass verification.
+8. The application present the key to the user and asks to add a new key to
+   the trusted database `DB` if the verification passes.
+9. If the user added the key as trusted, continue to boot. Otherwise go back
+   to the firmware setup or attempt the next boot option in order.
```

Comment:
What will happen if the application finds no suitable bootloader?

Reply:
Go to firmware setup.

Reply:
We should have it in flow.

Reply:
`Otherwise go back to the firmware setup` ??